src/OrleansConsulUtils/ConsulBasedMembershipTable.cs

```diff
@@ -33,46 +31,70 @@ public Boolean IsUpdatable
+            _maxStaleness = configuration.GatewayListRefreshPeriod;
+
+            this.deploymentId = configuration.DeploymentId;
+
+            consulClient =
+                new ConsulClient(config => config.Address = new Uri(configuration.DataConnectionString));
+
+            return Task.CompletedTask;
         }
 
-        public Task InitializeGatewayListProvider(ClientConfiguration config)
+        public async Task<IList<Uri>> GetGateways()
         {
-            _maxStaleness = config.GatewayListRefreshPeriod;
+            var membershipTableData = await ConsulBasedMembershipTable.ReadAll(this.consulClient, this.deploymentId, this.logger);
+            if (membershipTableData == null) return new List<Uri>();
+
+            return membershipTableData.Members.Select(e => e.Item1).
+                Where(m => m.Status == SiloStatus.Active && m.ProxyPort != 0).
+                Select(m =>
+                {
+                    m.SiloAddress.Endpoint.Port = m.ProxyPort;
+                    return m.SiloAddress.ToGatewayUri();
+                }).ToList();
+        }
+    }
 
-            Init(config.DeploymentId, config.DataConnectionString);
 
-            return Task.CompletedTask;
+    /// <summary>
+    /// A Membership Table implementation using Consul 0.6.0  https://consul.io/
+    /// </summary>
+    public class ConsulBasedMembershipTable : IMembershipTable
+    {
+        //Consul does not support the extended Membership Protocol and will always return the same table version information
+        public static readonly TableVersion _tableVersion = new TableVersion(0, "0");
+
+        private ILogger _logger;
+        private readonly ConsulClient _consulClient;
+        private readonly ConsulMembershipTableOptions membershipTableOptions;
+
+
+        public ConsulBasedMembershipTable(ILogger<ConsulBasedMembershipTable> logger,
+            IOptions<ConsulMembershipTableOptions> membershipTableOptions)
+        {
+            this._logger = logger;
+            this.membershipTableOptions = membershipTableOptions.Value;
+            _consulClient =
+                new ConsulClient(config => config.Address = new Uri(this.membershipTableOptions.DataConnectionString));
         }
 
         /// <summary>
         /// Initializes the Consul based membership table.
         /// </summary>
-        /// <param name="config">The configuration for this instance.</param>
         /// <param name="tryInitTableVersion">Will be ignored: Consul does not support the extended Membership Protocol TableVersion</param>
         /// <returns></returns>
         /// <remarks>
         /// Consul Membership Provider does not support the extended Membership Protocol,
         /// therefore there is no MembershipTable to Initialise
         /// </remarks>
-        public Task InitializeMembershipTable(GlobalConfiguration config, Boolean tryInitTableVersion)
+        public Task InitializeMembershipTable(Boolean tryInitTableVersion)
```

Comment:
Boolean vs. bool

Reply:
Boolean come from the public IMembershipTable interface . Are you suggesting we change the pubic IMembershipTable interface?

Reply:
`bool` is a C# alias for the `System.Boolean` type. They are equivalent, it's just that by convention we use `bool` everywhere. Not sure why we got `Boolean` there in the first place